### ISSUE TITLE
Methods ParallelOperation TryBegin and void TryReceiveNextMessage are uncoordinated

### ISSUE THREAD

**Issue by BBGONE** [rebus-org/Rebus#756]
In the Rebus 5 there's a sharp performance drop when the bus is used to receive messages from transport which uses async reads (as an SqlServer transport)
The Receive:
1) o.SetMaxParallelism(5) and MaxNumberOfWorkers: 5 (measure the receive performance - i have 51 msgs/ sec)
2) o.SetMaxParallelism(20) and MaxNumberOfWorkers: 1 (measure the receive performance - i have 738 msgs/ sec)
3) o.SetMaxParallelism(20) and MaxNumberOfWorkers: 5 (measure the receive performance - i have 568 msgs/ sec)
4) o.SetMaxParallelism(20) and MaxNumberOfWorkers: 20 (measure the receive performance - i have 688 msgs/ sec)

The problem is due to the fact that the  backoff and async message receive (the InMemoryTransport uses synchronous message receive) . The worker starts async receive, and does not wait for its completion, but instead waits on the backoff.  After that even if the task have been completed already, it still waits on the backoff (not knowning that it can start a new task). So most of the time it waits on backoff, and does not start a new task.

The solution:
 in the ParallelOperationsManager change to this code:
```
 /// <summary>
        /// Begins another async operation and returns an <see cref="IDisposable"/> that must be disposed in order to mark the end of the async operation
        /// </summary>
        public ParallelOperation TryBegin(CancellationToken  token)
        {
          // in the original code it is:       _semaphore.Wait(0);
           _semaphore.Wait(token);
            token.ThrowIfCancellationRequested();
            // always allow parallel operation because semaphore is free at this moment
            return new ParallelOperation(true, this);
        }
```
 in the ThreadPoolWorker to this code:

```
       void TryReceiveNextMessage(CancellationToken token)
        {
            var parallelOperation = _parallelOperationsManager.TryBegin(token);
            /*
            if (!parallelOperation.CanContinue())
            {
                _backoffStrategy.Wait(token);
                return;
            }
            */
            TryAsyncReceive(token, parallelOperation)
                .ContinueWith(LogException, TaskContinuationOptions.OnlyOnFaulted);
        }
```

The backoff should be used only to wait when no messages was received, not to wait for when parallel operation can continue - because it is better to use the semaphore for waiting for this moment

**Comment by BBGONE** [comment]
P.S. Although not related to this issue but i improved a bit FileSystemTransport (you could consider using it)
https://github.com/BBGONE/REBUS-TaskCoordinator

**Comment by BBGONE** [comment]
Oops, the suggested above solution adds another problem - the race conditions. To solve that it needs a lock around checking the condition (after semapore is released) and starting a new task. But you need to start the task explicitly because the transport as in memorytransport uses synchronous reads and you will have a lock around all the job (not only starting a new task).

But i suggest you to use my workerscoordinator introduced in the REBUS-TaskCoordinator project.
It solves all the problems and moves all the complexity of adding and removing tasks out of the bus. And i developed it a long time ago (the TasksCoordinator), so i went through all the bugs and mistakes in its implementation (and use in real world projects) and i suggest you not to step on the same rakes again.

**Comment by mookid8000** [comment]
Hi @BBGONE , thanks for taking the time to figure out these things.... I need more time and space in my head to process this information though, so please be patient with me 😁

**Comment by mookid8000** [comment]
hey @BBGONE , sorry for taking so long to get back to this one.... lots of stuff has happened since you reported this, both with Rebus and Rebus.SqlServer – if it's still relevant, please feel free to pick it up again 🙂
